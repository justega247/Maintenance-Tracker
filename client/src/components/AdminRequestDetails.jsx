import React, { Component } from "react";
import { bindActionCreators } from "redux";
import { connect } from "react-redux";
import { Link } from "react-router-dom";
import PropTypes from "prop-types";
import {
  startFetchRequest,
  startAdminRequestResolve,
  startAdminRequestDisapprove,
  startAdminRequestApprove
} from "../actions/request";
import Button from "./common/Button";

export class AdminRequestDetails extends Component {
  componentDidMount = () => {
    const { match, fetchRequest } = this.props;
    if (match && match.params.id) {
      fetchRequest(match.params.id);
    }
  };

  onApprove = () => {
    const { match, approveRequest } = this.props;
    if (match && match.params.id) {
      approveRequest(match.params.id);
    }
  };

  onDisapprove = () => {
    const { match, disapproveRequest } = this.props;
    if (match && match.params.id) {
      disapproveRequest(match.params.id);
    }
  };

  onResolve = () => {
    const { match, resolveRequest } = this.props;
    if (match && match.params.id) {
      resolveRequest(match.params.id);
    }
  };

  render() {
    const { userRequest } = this.props;
    const { id, title, type, status, description } = userRequest;

    return (
      <div className="view__div">
        <div className="view__request">
          <div className="req__item">
            <div className="req__value">id:</div>
            <div>{id}</div>
          </div>
          <div className="req__item">
            <div className="req__value">Title:</div>
            <div>{title}</div>
          </div>
          <div className="req__item">
            <div className="req__value">Type:</div>
            <div>{type}</div>
          </div>
          <div className="req__item">
            <div className="req__value">Status:</div>
            <div>{status}</div>
          </div>
          <div className="req__item">
            <div className="req__value">Description:</div>
            <div>{description}</div>
          </div>
          <div>
            <Button
              type="button"
              text="Approve"
              onClick={this.onApprove}
              className="admin_manage details"
            />
            <Button
              type="button"
              text="Disapprove"
              onClick={this.onDisapprove}
              className="admin_manage disapprove"
            />
            <Button
              type="button"
              text="Resolve"
              onClick={this.onResolve}
              className="admin_manage resolve"
            />
          </div>
        </div>
      </div>
    );
  }
}

AdminRequestDetails.defaultProps = {
  fetchRequest: () => {},
  userRequest: {},
  match: {}
};

AdminRequestDetails.propTypes = {
  fetchRequest: PropTypes.func,
  userRequest: PropTypes.instanceOf(Object),
  match: PropTypes.instanceOf(Object)
};

const mapStateToProps = state => ({
  userRequest: state.requests.request
});

const mapDispatchToProps = dispatch => ({
  fetchRequest: bindActionCreators(startFetchRequest, dispatch),
  approveRequest: bindActionCreators(startAdminRequestApprove, dispatch),
  disapproveRequest: bindActionCreators(startAdminRequestDisapprove, dispatch),
  resolveRequest: bindActionCreators(startAdminRequestResolve, dispatch)
});

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(AdminRequestDetails);
