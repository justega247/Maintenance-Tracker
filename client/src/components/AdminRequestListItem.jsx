import React from "react";
import { Link } from "react-router-dom";
import PropTypes from "prop-types";
import Button from "./common/Button";

const formatDate = date => {
  const day = new Date(date).getDate();
  const month = new Date(date).getMonth() + 1;
  const year = new Date(date).getFullYear();

  const createdAt = `${day}/${month}/${year}`;
  return createdAt;
};
const AdminRequestListItem = ({ request }) => (
  <tr>
    <td data-label="Requester">{request.requested_by}</td>
    <td data-label="Title">{request.title}</td>
    <td data-label="Type">{request.type}</td>
    <td data-label="Manage">
      <Link to={`/admin-view-request/${request.id}`}>
        <Button type="button" text="Manage" className="details" />
      </Link>
    </td>
    <td data-label="Created">{formatDate(request.created_at)}</td>
  </tr>
);

AdminRequestListItem.propTypes = {
  request: PropTypes.instanceOf(Object).isRequired
};

export default AdminRequestListItem;
