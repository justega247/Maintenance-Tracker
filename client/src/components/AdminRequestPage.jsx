import React from 'react';

export default function AdminRequestPage() {
  return (
    <div>
      This is from the Admin request page
    </div>
  );
}
