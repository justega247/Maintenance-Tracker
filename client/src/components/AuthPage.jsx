import React from 'react';

const AuthForm = () => (
  <div>
    This is from the Auth component
  </div>
);

export default AuthForm;
