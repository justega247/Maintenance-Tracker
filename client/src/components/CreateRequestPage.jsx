import React, { Component } from 'react';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import PropTypes from 'prop-types';
import RequestForm from './RequestForm';
import { startAddRequest } from '../actions/request';

export class CreateRequestPage extends Component {
  onSubmit = (request) => {
    const { startCreateRequest, history } = this.props;
    startCreateRequest(request, history);
  };

  render() {
    return (
      <div className="create__request">
        <div>
          <div>
            <h1 className="create__request-title">Add Request</h1>
          </div>
        </div>
        <div className="content-container">
          <RequestForm
            onSubmit={this.onSubmit}
          />
        </div>
      </div>
    );
  }
}

CreateRequestPage.defaultProps = {
  startCreateRequest: () => {},
};

CreateRequestPage.propTypes = {
  startCreateRequest: PropTypes.func,
  history: PropTypes.shape({}).isRequired,
};

const mapDispatchToProps = dispatch => ({
  startCreateRequest: bindActionCreators(startAddRequest, dispatch),
});

export default connect(undefined, mapDispatchToProps)(CreateRequestPage);
