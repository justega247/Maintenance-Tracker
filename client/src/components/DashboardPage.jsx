import React from 'react';

const Dashboard = () => (
  <div>
    This is a test right
  </div>
);

export default Dashboard;
