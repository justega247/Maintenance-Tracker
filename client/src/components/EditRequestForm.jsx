import React, { Component } from 'react';
import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import ValidateRequests from '../utils/requestValidation';
import Button from './common/Button';
import { startFetchRequest, startEditRequest } from '../actions/request';


export class EditRequestForm extends Component {
  state = {
    title: '',
    type: '',
    description: '',
    errors: {},
  }

  componentDidMount = () => {
    const { match, fetchRequest } = this.props;
    if (match && match.params.id) {
      fetchRequest(match.params.id)
        .then((payload) => {
          const { title, type, description } = payload.data.data.request;
          return this.setState(() => ({
            title,
            type,
            description,
          }));
        });
    }
  }

  onChange = (event) => {
    const { name, value } = event.target;
    this.setState({ [name]: value });
  };

  onSubmit = (event) => {
    event.preventDefault();

    const { errors, hasErrors } = ValidateRequests.requestDataValidation(this.state);
    this.setState(() => ({
      errors,
    }));
    if (!hasErrors) {
      const { title, type, description } = this.state;
      const { startUpdateRequest, history, match } = this.props;
      startUpdateRequest(match.params.id, { title, type, description }, history);
    }
  };

  render() {
    const { errors } = this.state;

    const types = [{ type: 'repairs' }, { type: 'maintenance' }];

    return (
      <div className="create__request">
        <div>
          <div>
            <h1 className="create__request-title">Edit Request</h1>
          </div>
        </div>
        <div className="content-container">
          <div className="request__form">
            <form onSubmit={this.onSubmit}>
              <div>
                <label htmlFor="type" className="label request__label">
              Type:
                  <select
                    name="type"
                    value={this.state.type}
                    onChange={this.onChange}
                    className="request__type"
                  >
                    {
                types.map(data => (
                  <option value={data.type} key={data.type}>
                    {data.type}
                  </option>
                ))
                }
                  </select>
                </label>
              </div>
              <div>
                <label htmlFor="title" className="label request__label">
              Title:
                  <div>
                    <input
                      name="title"
                      type="text"
                      placeholder="enter a title"
                      className="form__input"
                      value={this.state.title}
                      onChange={this.onChange}
                    />
                  </div>
                </label>
                {errors.title &&
                  errors.title.map(error => (
                    <div className="form__error" key={error}>
                      {error}
                    </div>
                  ))}
              </div>
              <div>
                <label htmlFor="description" className="label request__label">
              Description:
                  <div>
                    <textarea
                      name="description"
                      placeholder="enter a description"
                      className="form__input"
                      value={this.state.description}
                      onChange={this.onChange}
                    />
                  </div>
                </label>
                {errors.description &&
                  errors.description.map(error => (
                    <div className="form__error" key={error}>
                      {error}
                    </div>
                  ))}
              </div>
              <div>
                <Button type="submit" className="button--big" text="Edit Request" />
              </div>
            </form>
          </div>
        </div>
      </div>

    );
  }
}

EditRequestForm.defaultProps = {
  fetchRequest: () => {},
  startUpdateRequest: () => {},
  match: {},
};

EditRequestForm.propTypes = {
  fetchRequest: PropTypes.func,
  startUpdateRequest: PropTypes.func,
  history: PropTypes.shape({}).isRequired,
  match: PropTypes.instanceOf(Object),
};

const mapDispatchToProps = dispatch => ({
  fetchRequest: bindActionCreators(startFetchRequest, dispatch),
  startUpdateRequest: bindActionCreators(startEditRequest, dispatch),
});

export default connect(null, mapDispatchToProps)(EditRequestForm);

