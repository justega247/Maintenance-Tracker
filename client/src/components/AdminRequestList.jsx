import React from "react";
import { connect } from "react-redux";
import PropTypes from "prop-types";
import AdminRequestListItem from "./AdminRequestListItem";

export const AdminRequestList = ({ requests }) => (
  <table>
    <thead>
      <tr>
        <th>Requester</th>
        <th>Title</th>
        <th>Type</th>
        <th>Manage</th>
        <th>Created</th>
      </tr>
    </thead>
    <tbody>
      {requests.map(request => (
        <AdminRequestListItem key={request.id} request={request} />
      ))}
    </tbody>
  </table>
);

AdminRequestList.propTypes = {
  requests: PropTypes.instanceOf(Array).isRequired
};

const mapStateToProps = state => ({
  requests: state.requests.requests
});

export default connect(mapStateToProps)(AdminRequestList);
