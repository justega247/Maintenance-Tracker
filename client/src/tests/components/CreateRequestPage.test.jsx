import React from 'react';
import { shallow } from 'enzyme';
import { CreateRequestPage } from '../../components/CreateRequestPage';
import mockData from '../__mocks__/mockData';

let wrapper;
let props;
describe('CreateRequestPage', () => {
  beforeEach(() => {
    props = {
      startCreateRequest: jest.fn(),
      history: { push: jest.fn() },
    };
    wrapper = shallow(<CreateRequestPage {...props} />);
  });

  test('should render CreateRequestPage correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  test('should handle onSubmit', () => {
    const { request } = mockData;
    wrapper.find('RequestForm').prop('onSubmit')(request);
    expect(props.startCreateRequest).toHaveBeenCalledWith(request, props.history);
  });
});
