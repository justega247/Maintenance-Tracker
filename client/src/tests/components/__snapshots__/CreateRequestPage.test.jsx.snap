// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreateRequestPage should render CreateRequestPage correctly 1`] = `
<div
  className="create__request"
>
  <div>
    <div>
      <h1
        className="create__request-title"
      >
        Add Request
      </h1>
    </div>
  </div>
  <div
    className="content-container"
  >
    <RequestForm
      onSubmit={[Function]}
    />
  </div>
</div>
`;
